spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/ConvertingAndDelegatingMessageListenerAdapter.java

```diff
@@ -0,0 +1,143 @@
+	 * Construct an instance with the provided {@link MessageListener} and {@link Class}
+	 * as a desired type of {@link ConsumerRecord}'s value after conversion. Default value of
+	 * {@link MessageConverter} is used, which is {@link SimpleMessageConverter}.
+	 *
+	 * @param delegate the {@link MessageListener} to use when passing converted {@link ConsumerRecord} further.
+	 * @param desiredValueType the {@link Class} setting desired type of {@link ConsumerRecord}'s value.
+	 */
+	public ConvertingAndDelegatingMessageListenerAdapter(Object delegate, Class<V> desiredValueType) {
+		validateMessageListener(delegate);
+		Objects.requireNonNull(desiredValueType);
+
+		this.delegate = delegate;
+		this.desiredValueType = desiredValueType;
+
+		this.messageConverter = new SimpleMessageConverter();
+	}
+
+	/**
+	 * Construct an instance with the provided {@link MessageListener}, {@link MessageConverter} and {@link Class}
+	 * as a desired type of {@link ConsumerRecord}'s value after conversion.
+	 *
+	 * @param delegate the {@link MessageListener} to use when passing converted {@link ConsumerRecord} further.
+	 * @param messageConverter the {@link MessageConverter} to use for conversion.
+	 * @param desiredValueType the {@link Class} setting desired type of {@link ConsumerRecord}'s value.
+	 */
+	public ConvertingAndDelegatingMessageListenerAdapter(Object delegate, MessageConverter messageConverter, Class<V> desiredValueType) {
+		validateMessageListener(delegate);
+		Objects.requireNonNull(messageConverter);
+		Objects.requireNonNull(desiredValueType);
+
+		this.delegate = delegate;
+		this.messageConverter = messageConverter;
+		this.desiredValueType = desiredValueType;
+	}
+
+	private void validateMessageListener(Object messageListener) {
+		Objects.requireNonNull(messageListener);
+		if (!(messageListener instanceof MessageListener)) {
+			throw new IllegalArgumentException("Passed message listener must be of MessageListener type");
+		}
+	}
+
+	@Override
+	public void onMessage(ConsumerRecord<T, U> data, Acknowledgment acknowledgment, Consumer<?, ?> consumer) { // NOSONAR
+		ConsumerRecord<T, V> convertedConsumerRecord = convertConsumerRecord(data);
+		if (this.delegate instanceof AcknowledgingConsumerAwareMessageListener) {
+			((AcknowledgingConsumerAwareMessageListener<T, V>) this.delegate).onMessage(convertedConsumerRecord, acknowledgment, consumer);
+		}
+		else if (this.delegate instanceof ConsumerAwareMessageListener) {
+			((ConsumerAwareMessageListener<T, V>) this.delegate).onMessage(convertedConsumerRecord, consumer);
+		}
+		else if (this.delegate instanceof AcknowledgingMessageListener) {
+			((AcknowledgingMessageListener<T, V>) this.delegate).onMessage(convertedConsumerRecord, acknowledgment);
+		}
+		else if (this.delegate instanceof MessageListener) {
+			((MessageListener<T, V>) this.delegate).onMessage(convertedConsumerRecord);
+		}
+	}
+
+	private ConsumerRecord<T, V> convertConsumerRecord(ConsumerRecord<T, U> data) { // NOSONAR
```

Comment:
I don't think we care somehow about `T` & `U` in the logic of this class.
In the end what we only need is a `V` for the data type we would like to convert to.

Reply:
I tried to replace T and U with just ?, but then it turned out that I can't override onMessage method without compiler yelling at me with communicate "same erasure, yet neither overrides the other". Also from code readability perspective it's kinda useful in my opinion, because it's clearly visible, that key (T) stays the same and U is at the end converted to V. Anyway, if you don't feel convinced I'll try to put some additional effort to investigate how could I avoid using T and U keeping only V as a type of a value after conversion.